src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -807,63 +807,206 @@ ErrorIfInsertPartitionColumnDoesNotMatchSelect(Query *query, RangeTblEntry *inse
+		targetTableHasPartitionColumn = true;
+
+		subqeryTargetEntry = list_nth(subquery->targetList,
+									  insertVar->varattno - 1);
+
+		/*
+		 * Corresponding (i.e., in the same ordinal position as the target table's
+		 * partition column) select target entry does not directly belong a table.
+		 * Evaluate its expression type and error out properly.
+		 */
+		if (subqeryTargetEntry->resorigtbl == InvalidOid)
+		{
+			char *exprTypeDefault = "an expression that is not a plain column";
+			char *errorDetailTemplate = "Subquery contains %s in the "
+										"same position of the target table's "
+										"partition column.";
+
+			Expr *selectTargetExpr = subqeryTargetEntry->expr;
+			NodeTag exprType = selectTargetExpr->type;
+			char *errorDetail = "";
+			char *exprTypeStr = "";
 
-			subqeryTargetEntry = list_nth(subquery->targetList,
-										  insertVar->varattno - 1);
 
-			if (!IsA(subqeryTargetEntry->expr, Var))
+			switch (exprType)
 			{
-				partitionColumnsMatch = false;
+				case T_Const:
+				{
+					exprTypeStr = "a constant value";
+				}
 				break;
-			}
 
-			/*
-			 * Reference tables doesn't have a partition column, thus partition columns
-			 * cannot match at all.
-			 */
-			if (PartitionMethod(subqeryTargetEntry->resorigtbl) == DISTRIBUTE_BY_NONE)
-			{
-				partitionColumnsMatch = false;
+				case T_OpExpr:
+				{
+					exprTypeStr = "an operator";
+				}
 				break;
-			}
 
-			if (!IsPartitionColumnRecursive(subqeryTargetEntry->expr, subquery))
-			{
-				partitionColumnsMatch = false;
+				case T_FuncExpr:
+				{
+					FuncExpr *subqueryFunctionExpr =
+						(FuncExpr *) subqeryTargetEntry->expr;
+
+					switch (subqueryFunctionExpr->funcformat)
+					{
+						case COERCE_EXPLICIT_CALL:
+						{
+							exprTypeStr = "a function call";
+						}
+						break;
+
+						case COERCE_EXPLICIT_CAST:
+						{
+							exprTypeStr = "an explicit cast";
+						}
+						break;
+
+						case COERCE_IMPLICIT_CAST:
+						{
+							exprTypeStr = "an implicit cast";
+						}
+						break;
+
+						default:
+						{
+							exprTypeStr = "a function call";
+						}
+						break;
+					}
+				}
+				break;
+
+				case T_Aggref:
+				{
+					exprTypeStr = "an aggregation";
+				}
+				break;
+
+				case T_CaseExpr:
+				{
+					exprTypeStr = "a case expression";
+				}
+				break;
+
+				case T_CoalesceExpr:
+				{
+					exprTypeStr = "a coalesce expression";
+				}
+				break;
+
+				case T_RowExpr:
+				{
+					exprTypeStr = "a row expression";
+				}
+				break;
+
+				case T_MinMaxExpr:
+				{
+					exprTypeStr = "a min/max expression";
+				}
+
+				break;
+
+				case T_CoerceViaIO:
+				{
+					exprTypeStr = "an explicit coercion";
+				}
+
+				break;
+
+				default:
+				{
+					exprTypeStr = exprTypeDefault;
+				}
 				break;
 			}
 
-			partitionColumnsMatch = true;
-			*selectPartitionColumnTableId = subqeryTargetEntry->resorigtbl;
+			errorDetail = psprintf(errorDetailTemplate, exprTypeStr);
```

Comment:
Is there a reason this is necessary rather than just passing the `errorDetailTemplate` and `exprTypeString` to `errdetail`? Couldn't it format the string for us (you're already using a `%s` there, anyhow)?